deploy-service/common/src/main/java/com/pinterest/arcee/aws/EC2HostInfoDAOImpl.java

```diff
@@ -58,48 +57,35 @@ public EC2HostInfoDAOImpl(AmazonEC2Client client) {
         HashSet<String> terminatedHosts = new HashSet<>();
-        while (!ids.isEmpty()) {
-            DescribeInstancesRequest request = new DescribeInstancesRequest();
-            request.setInstanceIds(ids);
+
+        for (String staleId : staleIds) {
             try {
-                do {
-                    DescribeInstancesResult results = ec2Client.describeInstances(request);
-                    List<Reservation> reservations = results.getReservations();
-                    for (Reservation reservation : reservations) {
-                        for (Instance instance : reservation.getInstances()) {
-                            int stateCode = instance.getState().getCode();
-                            String id = instance.getInstanceId();
-                            if (stateCode == TERMINATED_CODE || stateCode == STOPPED_CODE) {
-                                LOG.info(String.format("Instance %s has already been terminated or stopped.", id));
-                                terminatedHosts.add(id);
-                            }
-
-                            ids.remove(id);
+                DescribeInstancesRequest request = new DescribeInstancesRequest();
+                request.setInstanceIds(Collections.singletonList(staleId));
+                DescribeInstancesResult result = ec2Client.describeInstances(request);
+                List<Reservation> reservations = result.getReservations();
+                for (Reservation reservation : reservations) {
+                    for (Instance instance : reservation.getInstances()) {
+                        int stateCode = instance.getState().getCode();
+                        String id = instance.getInstanceId();
+                        if (stateCode == TERMINATED_CODE || stateCode == STOPPED_CODE) {
+                            LOG.info(String.format(
+                                "Instance %s has already been terminated or stopped.", id));
+                            terminatedHosts.add(id);
                         }
                     }
-                    if (results.getNextToken() == null || results.getNextToken().isEmpty()) {
-                        break;
-                    }
-                    request = new DescribeInstancesRequest();
-                    request.setInstanceIds(ids);
-                    request.setNextToken(results.getNextToken());
-                } while (true);
-                LOG.debug("Cannot find the following ids in AWS:", ids);
-                terminatedHosts.addAll(ids);
-                return terminatedHosts;
+                }
             } catch (AmazonServiceException ex) {
-                Collection<String> invalidHostIds = handleInvalidInstanceId(ex);
-                ids.removeAll(invalidHostIds);
-                // add invalid host ids to the terminated host list.
-                terminatedHosts.addAll(invalidHostIds);
+                Collection<String> invalidIds = handleInvalidInstanceId(ex);
```

Comment:
in handleInvalidInstanceId(), we still parse the id in the exception, could we just assume the id in the exception is the staleId here? The idea of handling one instance a time is so that we could get rid of the potentially fragile parsing code; otherwise, I would think describe instances in a batch makes more sense